### PR TITLE
Timeout business logic implementation

### PR DESCRIPTION
**Pull request by dottorblaster**
~Still have to write tests, but~ this PR implements the business logic of the timeout handling.

### CODE REVIEW

File: lib/wanda/execution/agent_check_error.ex [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  @derive Jason.Encoder
+  defstruct [
+    :agent_id,
+    :error,
+    :message
+  ]
+
+  @type t :: %__MODULE__{
+          agent_id: String.t(),
+          error: atom(),
```
Comment (fabriziosestito) [inline lib/wanda/execution/agent_check_error.ex:15]: `type:  :timeout`

File: lib/wanda/execution/server.ex [hunk 2]
```diff
@@ -73,8 +73,22 @@ defmodule Wanda.Execution.Server do
...
-        %State{} = state
+        %State{
+          execution_id: execution_id,
+          group_id: group_id,
+          gathered_facts: gathered_facts,
+          targets: targets,
+          agents_gathered: agents_gathered
+        } = state
       ) do
+    timed_out_agents =
```
Comment (fabriziosestito) [inline lib/wanda/execution/server.ex:84]: we should refer to this as "targets'

File: lib/wanda/execution/evaluation.ex [hunk 3]
```diff
@@ -119,8 +119,12 @@ defmodule Wanda.Execution.Evaluation do
        ) do
     expectation_results =
       agents_check_results
-      |> Enum.flat_map(fn %AgentCheckResult{expectation_evaluations: expectation_evaluations} ->
-        expectation_evaluations
+      |> Enum.flat_map(fn
```
Comment (fabriziosestito) [inline lib/wanda/execution/evaluation.ex:122]: imho being explicit here is better than being implicit, I would `Enum.filter` `AgentCheckError` and then keep the flat map as it was

Comment (dottorblaster) [reply lib/wanda/execution/evaluation.ex:126]: :+1: done

File: lib/wanda/execution/server.ex [hunk 4]
```diff
@@ -80,9 +80,24 @@ defmodule Wanda.Execution.Server do
...
+        } = state
       ) do
-    :ok = Messaging.publish("results", :timeout)
+    targets =
+      Enum.filter(targets, fn %Target{agent_id: agent_id} ->
+        agent_id not in agents_gathered
+      end)
+
+    gathered_facts = Gathering.put_gathering_timeouts(gathered_facts, targets)
+    result = Evaluation.execute(execution_id, group_id, checks, gathered_facts)
```
Comment (arbulu89) [inline lib/wanda/execution/server.ex:98]: This will most probably fail. We fixed this using:
`execution_completed = Messaging.Mapper.to_execution_completed(result)` and publishing that variable

**Review by arbulu89 — COMMENTED** [review]
Just to open a discussion.
It makes me wonder if we should send a different message in this situation. We already can send different messages with protobuf, and in this case is strange that we need to fill all the expected checks with timeout, instead of sending something like `ExecutionTimeout` struct.

At the end, this timeout scenario is pretty specific -> we didn't get the facts so we couldn't run the checks. It is something generic, not check specific.

My 2 cents

**Comment by fabriziosestito** [comment]
> Just to open a discussion. It makes me wonder if we should send a different message in this situation. We already can send different messages with protobuf, and in this case is strange that we need to fill all the expected checks with timeout, instead of sending something like `ExecutionTimeout` struct.


> At the end, this timeout scenario is pretty specific -> we didn't get the facts so we couldn't run the checks. It is something generic, not check-specific.
> 
> My 2 cents


@arbulu89  the idea was to have a best-effort approach to the execution. `ExecutionTimeout` would not allow us to send an event which is partial.
We thought about sending a list of the agents who timed out, in a separate field, or even a `ExecutionCompletedWithTimeouts` event, maybe we can follow up to this.

File: lib/wanda/execution/server.ex (hunk 4 above)
Comment (dottorblaster) [reply lib/wanda/execution/server.ex:98]: Ah yeah you're right, my bad

File: lib/wanda/execution/evaluation.ex [hunk 5]
```diff
@@ -54,10 +55,14 @@ defmodule Wanda.Execution.Evaluation do
          agents_facts
        ) do
     agents_results =
-      Enum.map(agents_facts, fn {agent_id, facts} ->
-        %AgentCheckResult{agent_id: agent_id}
-        |> add_agent_expectation_result(facts, expectations)
-        |> add_facts(facts, check_id)
+      Enum.map(agents_facts, fn
+        {agent_id, :timeout} ->
+          %AgentCheckError{agent_id: agent_id, type: :timeout, message: "Timeout dude"}
```
[the remark this answers is not in the document]
Comment (dottorblaster) [reply lib/wanda/execution/evaluation.ex:60]: Actually half a leftover half an "I'm glad you noticed it so you can give me an advice about a better string" :smile:

Do you have any better idea? I really didn't know what to put in there :sweat_smile:

File: lib/wanda/execution/server.ex [hunk 6]
```diff
@@ -80,9 +80,24 @@ defmodule Wanda.Execution.Server do
...
+          agents_gathered: agents_gathered
+        } = state
       ) do
-    :ok = Messaging.publish("results", :timeout)
+    targets =
+      Enum.filter(targets, fn %Target{agent_id: agent_id} ->
+        agent_id not in agents_gathered
+      end)
+
+    gathered_facts = Gathering.put_gathering_timeouts(gathered_facts, targets)
```
[the remark this answers is not in the document]
Comment (dottorblaster) [reply lib/wanda/execution/server.ex:97]: Well actually here we did, because maybe we have some gathered facts and some timeouts. But I'm ok changing it

**Comment by dottorblaster** [comment]
@arbulu89 I think @fabriziosestito told you the big picture already :sweat_smile: yeah I'm absolutely open to add an outer field in a follow-up. Also this kind of infrastructure allows us to set up the field for more cases in which we have an error, and reuse basically the same pattern and the same functions if we find more corner cases. Thanks for the feedback!

File: test/execution/server_test.exs [hunk 7]
```diff
@@ -142,5 +142,30 @@ defmodule Wanda.Execution.ServerTest do
 
       assert_receive {:DOWN, ^ref, _, ^pid, :normal}
     end
+
+    test "should go down when the timeout function gets called" do
```
Comment (fabriziosestito) [inline test/execution/server_test.exs:146]: here we should test that the facts were partially gathered by sending some of the agents `receive_facts`

File: test/execution/server_test.exs [hunk 8]
```diff
@@ -142,5 +142,30 @@ defmodule Wanda.Execution.ServerTest do
...
+             config: [timeout: 99_999]
+           ]}
+        )
+
+      ref = Process.monitor(pid)
+
+      Process.send(pid, :timeout, [:noconnect])
+
+      assert_receive {:DOWN, ^ref, _, ^pid, :normal}
+    end
```
Comment (fabriziosestito) [inline test/execution/server_test.exs:169]: a test that proves that the timeout is invoked correctly after the timeout period is missing

**Review by fabriziosestito — CHANGES_REQUESTED** [review]

File: lib/wanda/execution/gathering.ex [hunk 9]
```diff
@@ -19,6 +19,19 @@ defmodule Wanda.Execution.Gathering do
     )
   end
 
+  @doc """
+  Adds timeout data to gathered facts.
+  """
+  @spec put_gathering_timeouts(map(), [Target.t()]) :: map()
+  def put_gathering_timeouts(gathered_facts, timed_out_agents) do
```
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply lib/wanda/execution/gathering.ex]: @arbulu89 @dottorblaster sorry guys this should be called `gathered_facts`, as it is the map that represents all the `gathered_facts`  in the GenServer. Also, the function above uses the same naming and we use the same naming in the Execution server to distinguish facts that have been gathered from facts in the catalog.

Could you please revert to the original naming?

File: lib/wanda/execution/server.ex (hunk 6 above)
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply lib/wanda/execution/server.ex:99]: same as above, the naming is correct

File: lib/wanda/execution/evaluation.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply lib/wanda/execution/evaluation.ex:61]: @dottorblaster  "Agent timed out during the execution"

**Review by fabriziosestito — APPROVED** [review]
only one thing to change then green light from my side

File: lib/wanda/execution/evaluation.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply lib/wanda/execution/evaluation.ex:61]: Yes, that message is better hehe

**Review by arbulu89 — APPROVED** [review]
Looks good!
